### PR TITLE
[nnc] Add hardsigmoid

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59405 Fix symbolic derivative of hardswish
* **#59069 [nnc] Add hardsigmoid**

Differential Revision: [D28738166](https://our.internmc.facebook.com/intern/diff/D28738166)

### CODE REVIEW

**Comment by bertmaher** [comment]
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28738166).

**Review by eellison — APPROVED** [review]
looks good! if you add symbolic script support it will also speed up training and not just inference :)

**Comment by bertmaher** [comment]
> looks good! if you add symbolic script support it will also speed up training and not just inference :)

Nice!  Do we have a test harness for backward where I can add hardsigmoid to make sure I've done my differentiation correctly?

**Comment by eellison** [comment]
Yep, you can follow as https://github.com/pytorch/pytorch/pull/58773/files

File: torch/csrc/jit/runtime/symbolic_script.cpp [hunk 1]
```diff
@@ -910,6 +910,12 @@ const std::vector<std::string> functions = {
 
             return result, backward
 
+        def hardsigmoid(self):
+            result = torch.hardsigmoid(self)
+            def backward(grad_output):
+                return torch.where((result > -3.) & (result < 3.), 0., grad_output * (1.0 / 6.0))
```
Comment (desertfire) [inline torch/csrc/jit/runtime/symbolic_script.cpp:916]: I think you need to swap the if and else part.

**Review by desertfire — CHANGES_REQUESTED** [review]
symbolic_script.cpp needs to updated.

**Comment by bertmaher** [comment]
Thank you.  I was misled by the order of args of Vec:vblend, and didn't think hard enough about the math :-(.  It's disturbing that no tests caught this?

**Comment by eellison** [comment]
yea, that is not good... I think we need to add more warmup runs to trigger the path

**Comment by Krovatkin** [comment]
@bertmaher could you try adding your test here : https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/jit_metaprogramming_utils.py I wonder if its the integration of autodiff testing into OpInfo or our actual autodiff testing that's borked.

**Review by desertfire — APPROVED** [review]

**Comment by bertmaher** [comment]
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28738166).
